src/compat.jl

```diff
@@ -0,0 +1,373 @@
+    nargs = length(e.args)
+
+    # `e` is an expr.  In many cases, it suffices to
+    # - guess that the kind name is the same as the expr head
+    # - add no syntax flags
+    # - map e.args to syntax tree children one-to-one
+    maybe_kind = find_kind(string(e.head))
+    st_k = isnothing(maybe_kind) ? K"None" : maybe_kind
+    st_flags = 0x0000
+    child_exprs = copy(e.args)
+
+    # The following are special cases where the kind, flags, or children are
+    # different from what we guessed above.
+    if Base.isoperator(e.head) && st_k === K"None"
+        # e.head is an updating assignment operator (+=, .-=, etc).  Non-=
+        # dotted ops are wrapped in a call, so we don't reach this.
+        s = string(e.head)
+        @assert s[end] === '=' && nargs === 2
+        if s[1] === '.'
+            st_k = K".op="
+            op = s[2:end-1]
+        else
+            st_k = K"op="
+            op = s[1:end-1]
+        end
+        child_exprs = [e.args[1], Symbol(op), e.args[2]]
+    elseif e.head === :comparison
+        for i = 2:2:length(child_exprs)
+            op = child_exprs[i]
+            @assert op isa Symbol
+            op_s = string(op)
+            if is_dotted_operator(op_s)
+                child_exprs[i] = Expr(:., Symbol(op_s[2:end]))
+            end
+        end
+    elseif e.head === :macrocall
+        @assert nargs >= 1
+        a1 = e.args[1]
+        child_exprs = collect_expr_parameters(e, 2)
+        if a1 isa Symbol
+            child_exprs[1] = Expr(:MacroName, a1)
+        elseif a1 isa Expr && a1.head === :(.) && a1.args[2] isa QuoteNode
+            child_exprs[1] = Expr(:(.), a1.args[1], Expr(:MacroName, a1.args[2].value))
+        elseif a1 isa GlobalRef
+            # syntax-introduced macrocalls
+            if a1.name === Symbol("@cmd")
+                # expr_children = []
+            elseif a1.name === Symbol("@doc")
+            elseif a1.name === Symbol("@int128_str")
+            elseif a1.name === Symbol("@int128_str")
+            elseif a1.name === Symbol("@big_str")
+            end
+        else
+            dump(e)
+            error("Unknown macrocall form $e")
+        end
+
+        # TODO node->expr handles do blocks here?
+    elseif e.head === Symbol("'")
+        @assert nargs === 1
+        st_k = K"call"
+        st_flags |= JS.POSTFIX_OP_FLAG
+        child_exprs = [e.head, e.args[1]]
+    elseif e.head === :. && nargs === 2
+        a2 = e.args[2]
+        if a2 isa Expr && a2.head === :tuple
+            st_k = K"dotcall"
+            tuple_exprs = collect_expr_parameters(a2, 1)
+            child_exprs = pushfirst!(tuple_exprs, e.args[1])
+        elseif a2 isa QuoteNode && a2.value isa Symbol
+            e.args[2] = a2.value
+        elseif a2 isa Expr && a2.head === :MacroName
+        else
+            @error "Unknown 2-arg dot form?" e
+        end
+    elseif e.head === :for
+        @assert nargs === 2
+        child_exprs = [_to_iterspec([e.args[1]]), e.args[2]]
```

Comment:
```suggestion
        child_exprs = Expr[_to_iterspec(Any[e.args[1]]), e.args[2]]
```

This should improve the inferred type of the inner vector + reduce specializations of `_to_iterspec`

The outer one is probably fine as-is, but it's pretty easy for inference to get it wrong, so I tend to prefer manually annotating the array type.

A similar tweak could be made in a few other places